### DISCUSSION TITLE
i am somewhat confused as to what pkl IS

### DISCUSSION THREAD

**Discussion by ghost** [apple/pkl#863] [Q&A]
so ive just heard about pkl and that it is a configuration language and am going through the tutorial,
however it seems that it is also a normal programming language (which reminds me more of lua which is used a extension/config lang)
and then the thing about language bindingd is also somewhat enigmatic to me.

ive not had much experience with this kind of language so please dont be too harsh

**Comment by bioball** [comment]
You're on the right track! It's _both_ a programming language and a configuration language.

Can you be more specific? I'm not sure what's confusing to you.

**Reply by ghost** [reply]
i think i was thinking of it wrong as for me a configuration is a rather static thing, as in not something you execute but rather something you read and write data from. i think im confusing it with a format. is me comparing it to lua sensible? (seing as its often used both to configure and extend programms), would the term extension language apply to pkl?

**Reply by ghost** [reply]
wait for it being an extension lang youd need to have something you can embedd into an application and i think that is hard with an interpreter which i think is written in java

**Reply by HT154** [reply]
The Lua comparison is both apt and not. You can use Pkl in many of the same ways as Lua but fundamentally they are very different: Lua is an imperative language where commands run in a programmer-specified order, while Pkl is a functional language that produces data as output (and order of execution is implicit and often unintuitive). Pkl is certainly not a general-purpose programming language like Lua is.

Pkl can be used in several ways:
* As a CLI tool to produce JSON/YAML/etc. static formats
* As a library directly from JVM languages (producing POJOs directly from evaluated Pkl code)
* Via a language binding library (eg. pkl-go or pkl-swift) similar in functionality to the JVM route, but communicating with the Pkl CLI over an RPC channel (the message passing API). Currently the Pkl CLI must be available to the application to actually evaluate Pkl code, but there are future plans to provide an alternate mechanism where Pkl is packaged as a library and linked directly to client applications.

The Pkl CLI itself is a native binary and does not depend on the JVM.

The closest comparison you can really make to Pkl are other config languages like cuelang and KCL: dynamic languages that produce data as output with a strong focus on the type system.

**Reply by ghost** [reply]
ok i think i get it.
so it is a language which when "executed" will produce the config values which can then be used by then main program?

**Reply by ghost** [reply]
would it make sense to compare pkl a bit to nix?

**Reply by HT154** [reply]
Yup, the Nix language is a pretty good comparison! Pkl's documentation calls out Jsonnet, HCL, and Dhall as other comparable languages: https://pkl-lang.org/main/current/introduction/comparison.html#other-config-langs

**Reply by ghost** [reply]
i think ill keep learning it then, afaik someone already made a lib for rust so that wont be an issue

**Comment by ghost** [comment]
from what i see most people saying about it people seem to see pkl as an intermediary language which will produce for example json files, can i assume that i can just use pkl directly? as in writing a program which just takes a pkl file as a config and the evaluates it internally to get the config data? or is that something it was never meant for?

**Reply by HT154** [reply]
Pkl can function in both ways! Both are common usages and well-supported.

If you control the application directly, it's often most ergonomic to accept Pkl directly as configuration. Libraries may also support code generation that—given a "schema" written in Pkl—generates types in the target language. The Pkl project's own libraries for Java, Kotlin, Go, and Swift support this.

If you don't control the application, you may need to generate config in a common format like JSON or YAML. Pkl is great at this as well. There are several built in renderers for different target languages and it's possible to write your own to produce any other textual format (eg. [TOML](https://github.com/apple/pkl-pantry/tree/main/packages/pkl.toml)). A great example of this approach is [pkl-k8s](https://github.com/apple/pkl-k8s), which produces YAML output for Kubernetes configurations.
